### DISCUSSION TITLE
It is possible to drop winston dependency in production?

### DISCUSSION THREAD

**Discussion by 7iomka** [UnlyEd/next-right-now#49] [Q&A] [answered]
Hi.
I enabled bundle analyze
with your implementation of logger 
`import { createLogger } from '@unly/utils-simple-logger';`
I need to install dependency `winston` to use it.
I only need logger in development.
It is possible to drop from production build?
Why?
![image](https://user-images.githubusercontent.com/18473137/82128331-ffd7be00-97c2-11ea-8a6e-6c22a55892cc.png)
![image](https://user-images.githubusercontent.com/18473137/82128341-154ce800-97c3-11ea-8adb-85a642c8137d.png)

**Comment by Vadorequest** [comment]
`logger` is also used in production, to report errors in the console.

But if you don't like it, just remove it then.

I don't know how to remove a package from build for a particular stage/env. I don't know how big Winston is (74kb, that much? Or is it 18.84kb?), I'll try to take a look at it at some point.

**Comment by 7iomka** [comment]
Without logger.
![image](https://user-images.githubusercontent.com/18473137/82133505-a9d03e00-97f5-11ea-99f8-a5229d7b9e66.png)

I find it too expensive to spend such resources for simply logging in production.
Please show your screenshot of the application build (I want to take a look for fun)

Now I’m just preparing myself an assembly for the start of the project, and I have borrowed your approaches regarding the organization of internationalization + multiversal-app-bootstrap, but it seems to me a very bad indicator that now these red numbers show when I have not written any business logic and the pages are almost completely empty!

**Comment by Vadorequest** [comment]
I agree with you, I wasn't aware it was so heavy.

Here is the production build:

![image](https://user-images.githubusercontent.com/3807458/82155019-446e6280-9872-11ea-8ac1-d84ab8239236.png)

Rewriting the logger will be a perf improvement. I wasn't so happy with it, but if it's that heavy then it's even worse than I thought. I'll try to find for a proper alternative, thanks for the heads-up!

**Comment by Vadorequest** [comment]
I'm considering https://github.com/klaussinani/signale as a replacement, much lighter (x10). I'll give it a try.

**Comment by Vadorequest** [comment]
signale didn't work out (as well as many others), eventually I tried ulog and made a PR https://github.com/UnlyEd/next-right-now/pull/50 but it's not usable yet.

It's crazy how hard it is to do something so simple...

**Comment by samuelcastro** [comment]
Yeah I've noticed that same issue, it would be great to find out a better alternative.

**Comment by Vadorequest** [comment]
If you know any, I'm interested, I've looked over ~6-10 of them so far. ulog and debug were my favorites so far.

**Comment by samuelcastro** [comment]
`debug` is more used and it's small, I've never used `ulog` though, looks interesting.

**Comment by 7iomka** [comment]
`ulog` really looks interesting

**Comment by Vadorequest** [comment]
Status of this:

I've attempted to use ulog, but couldn't make it work and didn't hear back from the author https://github.com/Download/ulog/issues/29

I don't have time to dig deeper, I agree that we could find a lighter alternative to Winston. (FYI I use https://bundlephobia.com/result?p=winston@3.2.1 to compare them before doing any work on them)

Feel free to propose a better alternative with its implementation. I'll leave the issue open until then.

**Comment by Vadorequest** [comment]
Gonna attempt with http://getpino.io/#/

---
Edit: Tried, didn't work. https://github.com/pinojs/pino/issues/840

**Comment by Vadorequest** [comment]
Considering after various attempts I didn't find anything that fits, what do you think about building a small internal utility for this?

The needs are:
- Shouldn't log stuff that isn't `error` in production (same behaviour on server and client)
- Ideally would log the name of the file that's calling
- Use the same API as the native `console` 
- Forwards calls so that we may know who is the initial caller when debugging (very important to understand where the initial `log()` comes from)
- Ideally avoid to configure one logger per file, it's just cumbersome
- Super light, no external dependency (ideally < 2ko)

Does those specs make sense?

Anyone would like to give it a try? I don't have the time 😕

**Comment by Vadorequest** [comment, marked as answer]
I finally wrote the above-mentioned utility and have started working on replacing the @unly/utils-simple-logger by a new @unly/simple-logger package.

The `winston` dependency has been removed, and been replaced by `chalk` (8kb gzipped) which is only being imported on the server.

The new package is <1Kb on the browser.

![image](https://user-images.githubusercontent.com/3807458/117549346-33e26980-b03a-11eb-9a9d-734f0c646839.png)


PR: https://github.com/UnlyEd/next-right-now/pull/314

**Reply by Vadorequest** [reply]
The PR has been merged!

[Lowered the size to <1kB](https://bundlephobia.com/result?p=@unly/simple-logger) 🎉
